File: algorithms/src/main/java/com/zlrx/algorithms/BreadthFirstSearch.java
package com.zlrx.algorithms;

import com.zlrx.algorithms.models.Vertex;

import java.util.LinkedList;
import java.util.Queue;

public class BreadthFirstSearch {

    /****************
     *       O      *
     *      / \     *
     *     4  3     *
     *    / \  \    *
     *   2  1  5    *
     ***************/
    public static void main(String[] args) {
        BreadthFirstSearch bfs = new BreadthFirstSearch();
        Vertex root = new Vertex(0);
        Vertex vertex1 = new Vertex(1);
        Vertex vertex2 = new Vertex(2);
        Vertex vertex3 = new Vertex(3);
        Vertex vertex4 = new Vertex(4);
        Vertex vertex5 = new Vertex(5);
        root.addNeighbour(vertex4);
        root.addNeighbour(vertex3);
        vertex4.addNeighbour(vertex2);
        vertex4.addNeighbour(vertex1);
        vertex3.addNeighbour(vertex5);
        bfs.bfs(root);
    }

    private void bfs(Vertex root) {
        Queue<Vertex> queue = new LinkedList<>();
        root.setVisited(true);
        queue.add(root);
        while (!queue.isEmpty()) {
            Vertex actual = queue.remove();
            System.out.println("Actual element is:" + actual);
            actual.getNeighbours().forEach(neighbour -> visitNeighbour(queue, neighbour));
        }
    }

    private void visitNeighbour(Queue<Vertex> queue, Vertex neighbour) {
        if (!neighbour.isVisited()) {
            neighbour.setVisited(true);
            queue.add(neighbour);
        }
    }

}

File: algorithms/src/main/kotlin/com/zlrx/algorithms/presentation/BreadthFirstSearch.kt
package com.zlrx.algorithms.presentation

import java.util.*


fun main(args: Array<String>) {
    BreadthFirstSearch().travel()
}

data class Node(val data: Int, val neighbours: List<Node> = listOf(), var visited: Boolean = false)

class BreadthFirstSearch {

    fun travel() {
        val root = createData()
        bfs(root)
    }

    fun createData(): Node {
        /****************
         *       O      *
         *      / \     *
         *     4  3     *
         *    / \  \    *
         *   2  1  5    *
         ***************/
        val node5 = Node(5)
        val node2 = Node(2)
        val node1 = Node(1)
        val node3 = Node(3, listOf(node5))
        val node4 = Node(4, listOf(node2, node1))
        val root = Node(0, listOf(node4, node3))
        return root
    }

    fun bfs(root: Node) {
        val queue = LinkedList<Node>()
        root.visited = true
        queue.add(root)

        while (!queue.isEmpty()) {
            val node = queue.remove()
            println("The actual node is $node")
            node.neighbours.forEach {
                if (!it.visited) {
                    it.visited = true
                    queue.add(it)
                }
            }
        }
    }


}







